### PR TITLE
Return all results and do not hang or crash when query with index returns null and not-null values [HZ-3012] [HZ-1842]

### PR DESCRIPTION
**Pull request by k-jamroz**
Fix ORDER BY queries with indexes which return both null and not-null values.

This PR fixes some of the issues uncovered in #25254. This is not the ultimate fix of all of them. To fix other issues `IndexIterationPointer` list has to be normalized i.e. converted to _ordered_ and _non-overlapping_ list of ranges. This will be done is subsequent PR and make some code from this PR a bit simpler.

Special NULL keys handling (`recordsWithNullValue`) is removed from on-heap sorted index implementation, similar to #25277 for HD. This makes `getSqlRecordIteratorBatch` NULL handling consistent for on-heap, HD and TS.

Fixes HZ-3012, https://github.com/hazelcast/hazelcast-enterprise/issues/5519
Follows #25254
EE PR: https://github.com/hazelcast/hazelcast-enterprise/pull/6415

Breaking changes (list specific methods/types/messages):
* Some SQL queries that previously returned incorrect results will now return different, correct results

Checklist:
- [x] Labels (`Team:`, `Type:`, `Source:`, `Module:`) and Milestone set
- [x] Label `Add to Release Notes` or `Not Release Notes content` set
- [x] Request reviewers if possible
- [ ] Send backports/forwardports if fix needs to be applied to past/future releases

### CODE REVIEW

File: hazelcast/src/main/java/com/hazelcast/query/impl/OrderedIndexStore.java [hunk 1]
```diff
@@ -193,51 +201,16 @@ public Iterator<IndexKeyEntries> getSqlRecordIteratorBatch(
                 = descending ? recordMap.descendingMap() : recordMap;
         switch (comparison) {
             case LESS:
-                if (descending) {
-                    navigableMap = navigableMap.tailMap(searchedValue, false);
-                } else {
-                    navigableMap = navigableMap.headMap(searchedValue, false);
-                }
-                break;
+                return getSqlRecordIteratorBatch(NULL, false, searchedValue, false, descending);
```
Comment (k-jamroz) [inline hazelcast/src/main/java/com/hazelcast/query/impl/OrderedIndexStore.java:204]: what was the rationale of having so many variants of `getSqlRecordIteratorBatch`? Is `tailMap`/`headMap` significantly cheaper than `subMap`? They all use the same implementation, but `null` end makes it slightly cheaper to find `loNode` and `hiNode`. Does it matter?

**Review by Fly-Style — COMMENTED** [review]
As a side note during the review: I'd propose to reduce quick `SqlIndexTest` just for integers.
The evaluation time right now is unacceptable for quick test.

**Comment by k-jamroz** [comment]
> As a side note during the review: I'd propose to reduce quick `SqlIndexTest` just for integers. The evaluation time right now is unacceptable for quick test.

Yeah, it is quite slow. I would keep also booleans because they have some specific tests (IS TRUE etc).

File: hazelcast/src/main/java/com/hazelcast/query/impl/IndexStore.java [hunk 2]
```diff
@@ -187,16 +194,25 @@ Iterator<QueryableEntry> getSqlRecordIterator(Comparable from, boolean fromInclu
     Iterator<IndexKeyEntries> getSqlRecordIteratorBatch(boolean descending);
 
     /**
+     * Returns records matching comparison. There are no records {@code < NULL}.
+     * Trying to get records {@code <= NULL} is invalid. However, you can get records {@code > NULL}
```
Comment (Fly-Style) [inline hazelcast/src/main/java/com/hazelcast/query/impl/IndexStore.java:198]: To be honest, ` records {@code < NULL}` and `records {@code <= NULL}` and are a bit confusing and hard to understand even after few re-readings. Can you re-word this doc, please?

**Review by Fly-Style — APPROVED** [review]
Well, after careful review the job done seems logical to me. Maybe @taburet will notice something in index-specific changes.
Nit: Enhance docs and provide more readable change description.

File: hazelcast/src/main/java/com/hazelcast/query/impl/IndexStore.java (hunk 2 above)
Comment (k-jamroz) [reply hazelcast/src/main/java/com/hazelcast/query/impl/IndexStore.java]: https://github.com/hazelcast/hazelcast/pull/25328/commits/9fb7383f38c58bc3aa4ae2e0c133d00e8bf0edcd

IMO this methods contract is hard to comprehend, but that is how it works. I only described it. These methods could be simplified but I do not understand why do we have so many variants of them.

File: hazelcast/src/main/java/com/hazelcast/map/impl/operation/MapFetchIndexOperation.java [hunk 3]
```diff
@@ -130,20 +122,43 @@ protected void runInternal() {
...
 
+    @Nonnull
+    public static InternalIndex getInternalIndex(@Nonnull MapContainer mapContainer,
+                                                 @Nonnull String mapName,
+                                                 @Nonnull String indexName) {
+        Indexes indexes = mapContainer.getIndexes();
+        if (indexes == null) {
+            throw QueryException.error(SqlErrorCode.INDEX_INVALID, "Cannot use the index \"" + indexName
+                    + "\" of the IMap \"" + mapName + "\" because it is not global "
+                    + "(make sure the property \"" + ClusterProperty.GLOBAL_HD_INDEX_ENABLED
```
Comment (petrpleshachkov) [inline hazelcast/src/main/java/com/hazelcast/map/impl/operation/MapFetchIndexOperation.java:133]: Do we have plans to support local indices here?

File: hazelcast/src/main/java/com/hazelcast/map/impl/operation/MapFetchIndexOperation.java [hunk 4]
```diff
@@ -130,20 +122,43 @@ protected void runInternal() {
         }
     }
 
+    @Nonnull
+    public static InternalIndex getInternalIndex(@Nonnull MapContainer mapContainer,
+                                                 @Nonnull String mapName,
+                                                 @Nonnull String indexName) {
+        Indexes indexes = mapContainer.getIndexes();
+        if (indexes == null) {
```
Comment (petrpleshachkov) [inline hazelcast/src/main/java/com/hazelcast/map/impl/operation/MapFetchIndexOperation.java:130]: You can use `mapContainer.isGlobalIndexEnabled()`

File: hazelcast/src/main/java/com/hazelcast/map/impl/operation/MapFetchIndexOperation.java (hunk 3 above)
Comment (k-jamroz) [reply hazelcast/src/main/java/com/hazelcast/map/impl/operation/MapFetchIndexOperation.java:133]: as part of https://hazelcast.atlassian.net/browse/HZ-2883 we probably will make this operation also working for local index (likely single partition at a time)

File: hazelcast/src/main/java/com/hazelcast/map/impl/operation/MapFetchIndexOperation.java (hunk 4 above)
Comment (k-jamroz) [reply hazelcast/src/main/java/com/hazelcast/map/impl/operation/MapFetchIndexOperation.java]: https://github.com/hazelcast/hazelcast/pull/25328/commits/45addd1c7ddb68a7df779f4ea84167154469b7c4, added also in some other places to more easily find spots where we need to add local index support

**Review by petrpleshachkov — APPROVED** [review]

**Pull request merged**